pyexchange/dydx.py

```diff
@@ -0,0 +1,261 @@
+
+    def __repr__(self):
+        return pformat(vars(self))
+
+    @staticmethod
+    def to_order(item: list, pair: str):
+        return Order(order_id=item['id'],
+                     timestamp=int(dateutil.parser.parse(item['createdAt']).timestamp()),
+                     pair=pair,
+                     is_sell=True if item['side'] == 'sell' else False,
+                     price=Wad.from_number(item['price']),
+                     amount=Wad.from_number(item['baseAmount']))
+
+
+class Trade:
+    def __init__(self,
+                 trade_id: str,
+                 timestamp: int,
+                 pair: str,
+                 price: Wad,
+                 amount: Wad):
+        assert (isinstance(trade_id, str) or (trade_id is None))
+        assert (isinstance(timestamp, int))
+        assert (isinstance(pair, str))
+        assert (isinstance(price, Wad))
+        assert (isinstance(amount, Wad))
+
+        self.trade_id = trade_id
+        self.timestamp = timestamp
+        self.pair = pair
+        self.price = price
+        self.amount = amount
+
+    def __eq__(self, other):
+        assert (isinstance(other, Trade))
+        return self.trade_id == other.trade_id and \
+               self.timestamp == other.timestamp and \
+               self.pair == other.pair and \
+               self.price == other.price and \
+               self.amount == other.amount
+
+    def __hash__(self):
+        return hash((self.trade_id,
+                     self.timestamp,
+                     self.pair,
+                     self.price,
+                     self.amount))
+
+    def __repr__(self):
+        return pformat(vars(self))
+
+    @staticmethod
+    def from_list(trade):
+        return Trade(trade_id=trade['uuid'],
+                     timestamp=int(dateutil.parser.parse(trade['createdAt']).timestamp()),
+                     pair=trade["market"],
+                     price=Wad.from_number(trade['price']),
+                     amount=Wad(int(trade['amount'])))
+
+
+class DydxApi(PyexAPI):
+    """Dydx API interface.
+
+        Documentation available here: https://docs.dydx.exchange/#/
+
+        Startup guide here: https://medium.com/dydxderivatives/programatic-trading-on-dydx-4c74b8e86d88
+    """
+
+    logger = logging.getLogger()
+
+    def __init__(self, node: str, private_key: str):
+        assert (isinstance(node, str))
+        assert (isinstance(private_key, str))
+
+        # self.address = web3.eth.accounts.privateKeyToAccount(privateKey)['address']
+        self.client = Client(private_key=private_key, node=node)
+
+    def get_markets(self):
+        return self.client.get_pairs()['pairs']
+
+    def get_pair(self, pair: str):
+        assert (isinstance(pair, str))
+        return next(filter(lambda symbol: symbol['name'] == pair, self.get_markets()))
+
+    def _balances_to_list(self, balances) -> List:
+        balance_list = []
+
+        for token, balance in enumerate(balances):
```

Comment:
can we reuse dydx constants as below e.g. ```consts.MARKET_ETH```

Reply:
what would be the purpose of this? This code is needed as the other services expect balance to provide as an array, not as a dictionary. I tried, but it wasn't clear how this could be made reusable so I didn't pull it out into a separate function.

Reply:
Nvm, I think i understand. I just updated the comparison to use consts instead of an int primitive as - e.g. `if token == consts.MARKET_ETH:`